ldms/src/sampler/cxi_sampler/cxi_sampler.c

```diff
@@ -0,0 +1,598 @@
+		cxi->tel_list_mval = ldms_metric_get(cxi->set, cxi->tel_list_mid);
+		rh_rec = ldms_record_alloc(cxi->set, cxi->rh_rec_mid);
+		tel_rec = ldms_record_alloc(cxi->set, cxi->tel_rec_mid);
+		rc = ldms_list_append_record(cxi->set, cxi->rh_list_mval, rh_rec);
+		if (rc) {
+			ovis_log(cxi->log, OVIS_LERROR,
+				 "Error %d appending record to rh_list.\n",
+				 rc);
+		}
+		rc = ldms_list_append_record(cxi->set, cxi->tel_list_mval, tel_rec);
+		if (rc) {
+			ovis_log(cxi->log, OVIS_LERROR,
+				 "Error %d appending record to tel_list.\n",
+				 rc);
+		}
+	}
+	return 0;
+
+ err:
+	if (cxi->base)
+		base_schema_delete(cxi->base);
+	return rc;
+}
+
+static const char *usage(ldmsd_plug_handle_t handle)
+{
+	static char help_str[PATH_MAX];
+	sprintf(help_str, "config name=%s %s [tel_path=PATH] [rh_path=PATH]",
+		ldmsd_plug_name_get(handle), BASE_CONFIG_USAGE);
+	return  help_str;
+}
+
+int get_cxi_interfaces(ldmsd_plug_handle_t handle, char *tel_path)
+{
+	cxi_t cxi = ldmsd_plug_ctxt_get(handle);
+	const char *pattern = "cxi*";
+	DIR *dir = opendir(tel_path);
+	int count;
+	if (!dir) {
+		ovis_log(cxi->log, OVIS_LERROR,
+			 "Error %d opening directory '%s'\n",
+			 errno, tel_path);
+		return 1;
+	}
+
+	struct dirent *ent;
+	count = 0;
+	while ((ent = readdir(dir)) != NULL) {
+		if (fnmatch(pattern, ent->d_name, 0) == 0) {
+			count++;
+		}
+	}
+	closedir(dir);
+	dir = opendir(tel_path);
+	cxi->iface_count = count;
+	cxi->iface_names = calloc(count, sizeof(char*));
+	if (!cxi->iface_names) {
+		ovis_log(cxi->log,
+			 OVIS_LERROR,
+			 "Memory allocation failed");
+		closedir(dir);
+		return 1;
+	}
+
+	count = 0;
+	while ((ent = readdir(dir)) != NULL) {
+		if (fnmatch(pattern, ent->d_name, 0) == 0) {
+			cxi->iface_names[count] = strdup(ent->d_name);
+			if (!cxi->iface_names[count]) {
+				ovis_log(cxi->log,
+					 OVIS_LERROR,
+					 "Memory allocation failed");
+				closedir(dir);
+				return 1;
+			}
+			count ++;
+		}
+	}
+	cxi->tel_files = calloc(count, sizeof(struct files_s));
+	if (!cxi->tel_files) {
+		ovis_log(cxi->log,
+			 OVIS_LERROR,
+			 "Memory allocation failed");
+		closedir(dir);
+		return 1;
+	}
+	cxi->rh_files = calloc(count, sizeof(struct files_s));
+	if (!cxi->rh_files) {
+		ovis_log(cxi->log,
+			 OVIS_LERROR,
+			 "Memory allocation failed");
+		closedir(dir);
+		return 1;
+	}
+	closedir(dir);
+	return 0;
+}
+
+static int config(ldmsd_plug_handle_t handle,
+		  struct attr_value_list *kwl, struct attr_value_list *avl)
+{
+	cxi_t cxi = ldmsd_plug_ctxt_get(handle);
+	ovis_log_t log = ldmsd_plug_log_get(handle);
+	int rc;
+	char *tel_path;
+	char *rh_path;
+
+	if (cxi->set) {
+		ovis_log(log, OVIS_LERROR, "Set already created.\n");
+		return EBUSY;
+	}
+
+	cxi->base = base_config(avl,
+				ldmsd_plug_cfg_name_get(handle),
+				ldmsd_plug_name_get(handle), log);
+	if (!cxi->base) {
+		rc = errno;
+		goto err;
+	}
+
+	tel_path = av_value(avl, "tel_path");
+	if (!tel_path) {
+		tel_path = "/sys/kernel/debug/cxi";
```

Comment:
Is there a reason to use this path instead of /sys/class/cxi? /sys/kernel/debug/cxi doesn't look like it works in older SHS releases. And libcxi still appears to use "/sys/class/cxi/%s/device/telemetry/%s" in the latest SHS release, so maybe that is the more universal one to use?

Reply:
It looks like @jstile-lbl agrees that it should be the /sys/class/cxi instead of /sys/kernel/debug/cxi.